### ISSUE TITLE
Support for history and location

### ISSUE THREAD

**Issue by thejohnhoffer** [WebReflection/linkedom#103]
Here's a CodeSandbox that shows parseHTML's document.defaultView doesn't contain `history` or `location`:

https://codesandbox.io/s/linkedom-no-history-o666v?file=/src/index.js

### Possible solutions:

When looking for `location` or `history`, I found [a comment](https://github.com/WebReflection/linkedom/issues/50#issuecomment-812931931) by @stephenh in issue #50:
> does linkedom support `window.location`?

I didn't see any folllow-up on that question, so I've used [remix-run/history][history] to simulate both `location` and `history` in [my own wrapper][wrapper] around `linkedom`.
[wrapper]: https://www.npmjs.com/package/linkedom-history

I'd be happy to work on a pull request to mock `history` and `location` within `linkedom`. Now, I use only one function called `createMemoryHistory` from "remix-run/history". Following linkedom's minimalist philosophy, I'd be happy to extract the `createMemoryHistory` function from "remix-run/history".

[history]: https://github.com/remix-run/history

**Comment by WebReflection** [comment]
Any JavaScript in this project has no meaning … there’s no JS here, if not DOM related. I don’t understand use cases around these requests but again, if it’s to simulate a browser, this is not the goal of this project.

if you would like to provide a lazy accessor for history or others API, as long as it’s lazy and doesn’t change anything else around this project, I’d be more than happy to have a look, but if your goal is to have JS running within linkedom documents, I recommend you go back to JSDOM or use cypress, puppeteer, or others solutions.

LinkeDOM is for SSR, not much else, really

**Comment by thejohnhoffer** [comment]
Hi @WebReflection -- thank you so much for this quick and clarifying response.

> if you would like to provide a lazy accessor for history or others API, as long as it’s lazy and doesn’t change anything else around this project, I’d be more than happy to have a look

Here's the quick pitch: if the user already has `remix-run/history` installed, then the `defaultView` would proxy `history` and `location` requests to a `MemoryHistory` from that library. 

> but if your goal is to have JS running within linkedom documents, I recommend you go back to JSDOM or use cypress, puppeteer, or others solutions.

I appreciate the guidance. With that in mind, it seems suprising how well `linkedom` works with `@testing-library/react`. I'd be interested to look deeper into why this `querySelector` call works if `linkedom` wasn't intended to work this way:

```jsx
import { parseHTML } from "linkedom"; 
import { render } from "@testing-library/react";  
const { document } = parseHTML(`<body><main></main></body>`); 
const container = document.querySelector("main")
render(<div>Custom JSX</div>, { container });
```

> LinkeDOM is for SSR, not much else, really

This certainly clears things up--- I've been using it for testing, and I suppose that's not recommended.

**Comment by thejohnhoffer** [comment]
I should clarify that `<div>Custom JSX</div>` in my example above could be replaced with a React Component that uses `react-router-dom` and therefore needs access to `window.history` or `document.defaultView.history`.

**Comment by WebReflection** [comment]
> I'd be interested to look deeper into why this querySelector call works if linkedom wasn't intended to work this way

selectors work because Custom Elements work too, builtin extends included, and there are other parts like MutationObserver that also work, but the main goal is still to SSR via workers, yet I am using linkedom myself for tests and code coverage, replacing basicHTML as it's too old and clunky compared to linkedom.

So yes, there are various scenarios where linkedom can work to test but I want the line to be clear: it's not a goal of this project to replace JSDOM or what a browser does.

If there is some quick/easy win that doesn't compromise performance or overcomplicate core logic I am OK, specially if lazy, but adding dependencies to test React only, as example, won't be cool ... so, how about proposing a way to enrich the global `window` / `defaultView` before parsing instead, so that anyone could be happy?

Because otherwise everyone wants their little extra bit on top, like [localStorage or window.onload](https://github.com/WebReflection/linkedom/issues/102) which is a complete non-sense to me.

**Comment by WebReflection** [comment]
P.S. the goal of this project is [defined in the README F.A.Q.s](https://github.com/WebReflection/linkedom#faq), under the first detail element ... maybe I should put that on top so it's more visible ...

**Comment by thejohnhoffer** [comment]
> If there is some quick/easy win that doesn't compromise performance or overcomplicate core logic I am OK

Thank you, I'll look into how this may be possible. I'll definitely leave the React stuff out of scope.

> P.S. the goal of this project is [defined in the README F.A.Q.s](https://github.com/WebReflection/linkedom#faq), under the first detail element ... maybe I should put that on top so it's more visible ...

Months ago I read the "not too close" section, but I only read the [contributing guidelines](https://github.com/WebReflection/linkedom/blob/main/how-to-contribute.md) before opening this issue. Maybe the guidelines could reference that section of the README?

**Comment by thejohnhoffer** [comment]
[This line](https://github.com/WebReflection/linkedom/blob/main/esm/interface/document.js#L132) of `interface/document.js` led me to realize that the `Document` __already proxies__ any global property to the `globalThis`. So, I can externally define `history` and `location` in the global object, and that will be reflected in the `document.defaultView` of `parseHTML`.

Maybe it would be cool to allow users to define their own additions to `globalExports` to avoid polluting the `globalThis` namespace, but certainly the global namespace is fine for my test cases.

**Comment by WebReflection** [comment]
> Maybe it would be cool to allow users to define their own additions to globalExports to avoid polluting the globalThis namespace, but certainly the global namespace is fine for my test cases.

Yes, and that's exactly what I had in mind when I said:

> how about proposing a way to enrich the global window / defaultView before parsing instead, so that anyone could be happy?

I didn't mean just pollute the `globalThis` with every possible unrelated trash, I meant a way to define, while parsing, or right before, any extra utility .. the returned global proxy might have a way to provide that, as opposite of polluting the `global` for everything not in there, I hope it's a bit more clear now what I meant

**Comment by thejohnhoffer** [comment]
> I hope it's a bit more clear now what I meant

Yes, thanks for bringing me up to speed.

Could this manifest as...
- a new `globalExports` parameter to [`parseFromString`](https://github.com/thejohnhoffer/linkedom/blob/b9774d6ff4c32c64aab092c1c3acb59e1b49ce2f/cjs/dom/parser.js#L21) and subsequently to the `HTMLDocument` constructor? 
- or perhaps an option for a custom user-defined `HTMLDocument` class (or proxy)?

**Comment by WebReflection** [comment]
my thinking, as `Symbol` is widely used in here, is that linkedom could export a special symbol that is considered in the global proxy too ... so you pass that symbol in the `parsefromString` or any other way to config your "*standard API call*", and if such Symbol, which can't conflict ever with anything standard, is present, its properties are returned with the Proxy ...

```js
parseFromString(html, {
  [LinkeDOMSpecialSymbol]: {
    get history() {},
    get location() {},
    anythingElse() { return 'really'; }
  }
})
```

so that the internal proxy, when this happens, passes along those things ... or something similar ... does it make sense?

**Comment by thejohnhoffer** [comment]
Yes, that makes sense. Would the second argument to parseFromString need to only contain the special Symbol as a key? Perhaps just that symbol and the current mimetype specification?

**Comment by WebReflection** [comment]
Any standard value *plus* eventually the special symbol ... it's an idea, not necessarily the best one.

**Comment by thejohnhoffer** [comment]
In my new PR #104, calling `parseFromString` with this new API copies any user defined keys/setters/getters to a proxy around `document.defaultView`, `document.defaultView.window`.
 
```js
const defaultView = parseFromString(html, {
  [linkedom.WINDOW]: {
    get thing() { return 'thing'; }
    anything() { return 'really'; }
  }
})
```

The old way of `parseFromString(html, "text/html")` still works, as well as a hybrid approach.

**Comment by WebReflection** [comment]
both `parseHTML(text, globals = null)` and `parseFromtString(text, mime, globals = null)` are now supported, so you can pass directly a global object that will be used to retrieve `location` or literally whatever you want.

I hope this is enough. Probably I am missing some reflection through the document, but the `window` has been tested and it works as expected.
